correct length and thisind for String

bkamins opened the pull request:
Following discussion in #25525 (this PR should be merged before working on #25525).

I do not touch `length` for `SubString` although it is broken now, because after #25525 it should work correctly as is.

bkamins commented:
I have added additional randomized tests. Additionally I had to fix:
* `thisind` because it referenced memory outside `s` sometimes which could cause problems;
* total rewrite of `_length` was required (it would be good if someone reviewed it carefully as the implementation it is complex as I wanted to get maximum speed - if we do not care about speed much most of the code can be replaced by calling `nextind` in `while` loop, but this should have small overhead);

bkamins commented:
I think CI failures are unrelated. This PR is ready for a review. Changes:
* fix `thisind` so that it does not access out of range data (bug hidden by `@inbounds`)
* fix `length` to correctly count number of valid characters by traversing the string like `nextind` (but optimized for speed). This is tricky (at least for me).

Merging this PR enables further work on:
* #25531 (correcting `nextind`/`prevind` for `SubString`)
* #25525 (fixing `convert` for `SubString`, improving and making consistent `getindex` for `String` and constructors of `SubString`)

StefanKarpinski commented:
I'm surprised that `_length` needs such an extensive rewrite. I don't understand what about our conversation in #25525 means that the core of the `String` `length` function needs to be completely rewritten. Can you explain what the issue is? It seemed like mostly only bounds checking for some cases needed to change.

Have you benchmarked these? The version that's there was carefully written to be fast for mostly-ASCII text (a common case). This case is fast because it spends all its in this inner loop:

https://github.com/JuliaLang/julia/blob/5c3f58039525972b24930f356821af8299f70a26/base/strings/string.jl#L258-L262

This inner loop has the following machine code on my system:
```asm
L48:
	movq	%rsi, %rdx
	movl	%r11d, %ecx
	addb	$64, %cl
	movzbl	8(%rdi,%r9), %r10d
	cmpb	$55, %cl
	jbe	L96            ; decode non-ASCII
	leaq	1(%rdx), %rsi
	movq	%rdx, %r9
	movl	%r10d, %r11d
	cmpq	%r8, %rsi
	jle	L48            ; continue ASCII loop
	jmp	L225           ; return
```
In order to accomplish this, `_length` is somewhat unusually structured – I would normally have written a single loop to process each character instead of exiting an inner loop to continue an outer loop like that. This allows the CPU branch prediction to perform really well for mostly ASCII text.

bkamins commented:
To tell the truth primarily I have rewritten it because I had problem getting the border cases right and later have forgotten to reinsert this inner `while` (it is clear why it is needed) - sorry for the confusion.

I do not insist on my implementation (and can try to fix-up yours) but maybe you would be convinced to my approach (let us just wait to make sure that it passes CI - I have added additional tests). See the benchmarks below (and maybe there is still a room for improvement of performance? - you have more experience here than me 😄).

The rewrite I have committed now actually speeds up the execution for large strings (it is a bit slower for very small strings so I have unrolled the loop for most critical cases) *both for ASCII only strings and for multibyte encoded strings*, it is only slower for long malformed strings (which I guess is not the target to optimize against).

Here are the benchmarks, where `length` is your current implementation and `xlength` is mine:
```
function f() # tests short strings
    for j in 1:2:11
        s = randstring(j)
        println((s, j))                 
        @time for i in 1:100_000_000 length(s) end 
        @time for i in 1:100_000_000 xlength(s) end
        s = randstring("⌣α∀", j)
        println((s, j))                 
        @time for i in 1:100_000_000 length(s) end 
        @time for i in 1:100_000_000 xlength(s) end
        s = String(rand(UInt8, j))
        println((s, j))                 
        @time for i in 1:100_000_000 length(s) end 
        @time for i in 1:100_000_000 xlength(s) end
    end
end                                 

function g() # tests long strings
    s = randstring(10^7)
    println("ASCII")
    @time for i in 1:10 length(s) end
    @time for i in 1:10 xlength(s) end
    println("mixed")
    s = randstring("aα∀", 10^7)
    @time for i in 1:10 length(s) end
    @time for i in 1:10 xlength(s) end
    println("long only")
    s = randstring("⌣α∀", 10^7)
    @time for i in 1:10 length(s) end
    @time for i in 1:10 xlength(s) end
    println("with invalid")
    s = String(rand(UInt8, 10^7))
    @time for i in 1:10 length(s) end
    @time for i in 1:10 xlength(s) end
end
```

And the results are:
```
julia> f()
("r", 1)
  0.280146 seconds
  0.078933 seconds
("α", 1)
  0.517611 seconds
  0.078778 seconds
("@", 1)
  0.281762 seconds
  0.080371 seconds
("sZE", 3)
  0.558846 seconds
  0.565410 seconds
("αα⌣", 3)
  1.414912 seconds
  1.137061 seconds
("O}\xe4", 3)
  0.533495 seconds
  0.607477 seconds
("RmZH3", 5)
  0.738497 seconds
  0.799457 seconds
("αα⌣∀⌣", 5)
  2.344901 seconds
  1.980074 seconds
("xƞ\x7f\xc2", 5)
  0.981383 seconds
  0.894113 seconds
("yXTk2zm", 7)
  1.008173 seconds
  0.962398 seconds
("∀⌣α⌣⌣α⌣", 7)
  3.273767 seconds
  2.934249 seconds
("<\xfc\x15\x81\x94\x0f/", 7)
  0.976811 seconds
  0.957999 seconds
("fLrfYQlJ8", 9)
  1.221775 seconds
  1.138996 seconds
("⌣α⌣⌣∀∀∀∀α", 9)
  4.359828 seconds
  3.631786 seconds
("\x1d'\x12\xee+@\x1c\x1dI", 9)
  1.466913 seconds
  1.244823 seconds
("C6ROsAe8oIE", 11)
  1.481095 seconds
  1.303011 seconds
("∀∀∀⌣⌣α⌣αα∀⌣", 11)
  5.127905 seconds
  4.318858 seconds
("\x86\x81\x0fo\x96j\x8b\xfb霅", 11)
  1.584240 seconds
  1.259733 seconds

julia> g()
ASCII
  0.095159 seconds
  0.061891 seconds
mixed
  0.864547 seconds
  0.851861 seconds
long only
  0.749923 seconds
  0.671426 seconds
with invalid
  0.332117 seconds
  0.378352 seconds
```

bkamins committed:
correct length and thisind for String

bkamins commented:
rebased

bkamins commented:
Although the tests pass there is a bug in the code. I will correct it later (and in the end maybe I will recommend to use the original implementation with fixed boundary checks).

bkamins commented:
Actually the fix has to be minimal using your code, so I have reverted to it to minimize changes.

As a side issue: the bug I have made (and similar bug fixed here) was caused by incorrect use of `@inbounds`. Maybe CI should use always use `check-bounds=yes` to catch such problems (but see #25566 that it seems that this flag does not always work).

bkamins commented:
CI errors seem unrelated.

bkamins commented:
Is this PR OK to be merged?

StefanKarpinski commented on base/strings/string.jl:
```diff
@@ -245,9 +246,8 @@ function length(s::String, i::Int, j::Int)
-    c = j - i + 1
     @inbounds i, k = thisind(s, i), i
-    c -= i < k
+    c = j - i + (i == k)
```
Does this change actually change the behavior? Or is this just an optimization?

bkamins replied:
Actually this is significant. Current (before this PR) behavior is:
```
julia> length("∀", 2, 3)
-1
julia> length("∀∀", 3, 4)
-1
```
this change makes sure that after this PR:
```
julia> length("∀", 2, 3)
0
julia> nlength("∀∀", 3, 4)
1
```

I made this change to avoid rewriting `_length` as you have suggested.

The meaning of this change is that we take `i` *after* aligning it with `thisind` not before it as it was done earlier - this caused the *decrementing* logic in `_length` to decrement too much sometimes (`c` was to small when `i` passed to `length` was inside multy-byte character). This caused a bug as shown in the above examples.

StefanKarpinski replied:
Got it. We can merge this as soon as CI passes.

StefanKarpinski reviewed (approved):
One question. Otherwise looks good to me.

bkamins commented:
CI errors seem unrelated.
